Handle leading whitespace on code blocks

bobbypriambodo opened the pull request:
Tries to address #133.

The implementation brings in `Str` module to split strings (since `split_on_chars` is available only since 4.04.0). Let me know if you have concerns.

I also wondered whether we need to specially handle weird spaces and tabs combo, but decided not to since I don't think there's a use case for that.

aantron commented on src/parser/lexer.mll:
```diff
@@ -70,6 +70,32 @@ let trim_trailing_blank_lines : string -> string = fun s ->
+  let least_amount_of_whitespace =
+    lines
+    |> List.map count_leading_whitespace
+    |> List.fold_left min max_int
```
I'd add a comment, that using `max_int` is okay because if the list is empty, we will not use the result of the `fold` (in the `map` below).

bobbypriambodo commented:
Thanks for the review. I went with the astring approach since it touches less files :D I've also added it on the `depends` field on the opam file.

Thanks for the catching on `max_int`, but it makes me wonder if the fact that it needed a comment means it's a bit of a smell. Unfortunately I can't think of a more straightforward implementation now, so I just added the note anyway.

Do tell if you have more concerns, would be happy to address them.

aantron commented:
Thanks for all that, it looks good to me :)

The `max_int` is a little bit of a smell, I guess, but I also can't think of something better.

aantron commented:
@bobbypriambodo There is one last thing remaining. I think we need to filter out empty or all-whitespace lines from computing the minimum:

```
{[
let () =
  Lwt_main.run begin
    let%lwt data = Lwt_io.(read_line stdin) in
    let%lwt () = Lwt_io.printl data in
    Lwt.return ()
  end

(* ocamlfind opt -linkpkg -package lwt.ppx,lwt.unix echo.ml && ./a.out *)
]}
```

In that example, the blank line is completely empty, so even when I indent the first and last line by two spaces, the blank line, which has no characters, forces the overall amount of leading whitespace to be computed as zero. Blank lines could have whitespace characters in general, though, even though that's sloppy coding :)

Of course, after ignoring the blank line, we will get two for the amount of leading space, so we have to be prepared for the blank line having less characters than that to substring out.

Let me know if you'd like me to fix this, but a quick follow-on PR is also welcome :)

bobbypriambodo commented:
Ah, that is true, I already tried it on several examples but none had that characteristics. Sorry for that. I think I'll be able to fix it tomorrow (it's past 11 PM on this part of the world). If that's not quick enough then please handle it as you see fit. Thanks.

aantron commented:
Thanks, that is fine :) and sorry for not catching right away.
